src/registrar/models/domain.py

```diff
@@ -300,3 +346,222 @@ def __str__(self) -> str:
+                        name=self.name,
+                        registrant=registrant.id,
+                        auth_info=epp.DomainAuthInfo(
+                            pw="2fooBAR123fooBaz"
+                        ),  # not a password
+                    )
+                    registry.send(req, cleaned=True)
+                    # no error, so go ahead and update state
+                    self.state = Domain.State.CREATED
+                    self.save()
+                else:
+                    raise e
+
+    def _get_or_create_contact(self, contact: PublicContact):
+        """Try to fetch info about a contact. Create it if it does not exist."""
+        while True:
+            try:
+                req = commands.InfoContact(id=contact.registry_id)
+                return registry.send(req, cleaned=True).res_data[0]
+            except RegistryError as e:
+                if e.code == ErrorCode.OBJECT_DOES_NOT_EXIST:
+                    create = commands.CreateContact(
+                        id=contact.registry_id,
+                        postal_info=epp.PostalInfo(  # type: ignore
+                            name=contact.name,
+                            addr=epp.ContactAddr(
+                                street=[
+                                    getattr(contact, street)
+                                    for street in ["street1", "street2", "street3"]
+                                    if hasattr(contact, street)
+                                ],
+                                city=contact.city,
+                                pc=contact.pc,
+                                cc=contact.cc,
+                                sp=contact.sp,
+                            ),
+                            org=contact.org,
+                            type="loc",
+                        ),
+                        email=contact.email,
+                        voice=contact.voice,
+                        fax=contact.fax,
+                        auth_info=epp.ContactAuthInfo(pw="2fooBAR123fooBaz"),
+                    )
+                    # security contacts should only show email addresses, for now
+                    if (
+                        contact.contact_type
+                        == PublicContact.ContactTypeChoices.SECURITY
+                    ):
+                        DF = epp.DiscloseField
+                        create.disclose = epp.Disclose(
+                            flag=False,
+                            fields={DF.FAX, DF.VOICE, DF.ADDR},
+                            types={DF.ADDR: "loc"},
+                        )
+                    registry.send(create)
+                else:
+                    raise e
+
+    def _update_or_create_host(self, host):
+        raise NotImplementedError()
+
+    def _delete_host(self, host):
+        raise NotImplementedError()
+
+    def _fetch_cache(self, fetch_hosts=False, fetch_contacts=False):
```

Comment:
**comment (non-blocking)**: This is a very large method. Is it possible that we might break it down into separate `_fetch_info`, `_fetch_contacts`, `_fetch_hosts` methods that get called from in here?